File: web/src/lib/mermaid.ts
const MERMAID_LINE_BREAK_RE = /<br\s*\/?>/gi;
const MERMAID_HTML_TAG_RE = /<\/?[A-Za-z][^>\n]*>/g;
const MERMAID_GRAPH_RE = /^(\s*)graph(\s+(?:TB|TD|BT|RL|LR)\b)/i;
const MERMAID_RECT_NODE_RE = /(\b[A-Za-z_][\w-]*)\[([^\[\]\n"]+)\]/g;
const MERMAID_SKIP_PREFIXES = [
	'%%',
	'subgraph ',
	'classDef ',
	'class ',
	'style ',
	'linkStyle ',
	'click ',
	'accTitle:',
	'accDescr:',
	'direction ',
];

const shouldSkipMermaidLine = (line: string) => {
	return line === 'end' || MERMAID_SKIP_PREFIXES.some((prefix) => line.startsWith(prefix));
};

const normalizeMermaidLabel = (label: string) => {
	return label
		.replace(MERMAID_LINE_BREAK_RE, ' ')
		.replace(MERMAID_HTML_TAG_RE, ' ')
		.replaceAll('&nbsp;', ' ')
		.replace(/\s+/g, ' ')
		.trim()
		.replaceAll('\\', '\\\\')
		.replaceAll('"', '\\"');
};

export const normalizeMermaidDiagram = (diagram: string) => {
	let normalized = diagram.replace(/\r\n/g, '\n').replace(/\r/g, '\n').trim();

	normalized = normalized.replace(MERMAID_LINE_BREAK_RE, ' ');
	normalized = normalized.replaceAll('&nbsp;', ' ');
	normalized = normalized.replace(MERMAID_GRAPH_RE, '$1flowchart$2');

	return normalized
		.split('\n')
		.map((line) => {
			const trimmed = line.trimStart();
			if (shouldSkipMermaidLine(trimmed)) {
				return line.trimEnd();
			}

			return line.replace(MERMAID_RECT_NODE_RE, (_, id: string, label: string) => {
				return `${id}["${normalizeMermaidLabel(label)}"]`;
			}).trimEnd();
		})
		.join('\n')
		.trim();
};

File: app/common/mermaid.py
import re


_MERMAID_FENCE_RE = re.compile(r"```mermaid[ \t]*\n(.*?)```", re.IGNORECASE | re.DOTALL)
_MERMAID_LINE_BREAK_RE = re.compile(r"<br\s*/?>", re.IGNORECASE)
_MERMAID_HTML_TAG_RE = re.compile(r"</?[A-Za-z][^>\n]*>")
_MERMAID_GRAPH_RE = re.compile(r"^(\s*)graph(\s+(?:TB|TD|BT|RL|LR)\b)", re.IGNORECASE)
_MERMAID_RECT_NODE_RE = re.compile(r'(?P<id>\b[A-Za-z_][\w-]*)\[(?P<label>[^\[\]\n"]+)\]')

_MERMAID_LINE_PREFIX_SKIP = (
    "%%",
    "subgraph ",
    "classDef ",
    "class ",
    "style ",
    "linkStyle ",
    "click ",
    "accTitle:",
    "accDescr:",
    "direction ",
)


def _should_skip_mermaid_line(line: str) -> bool:
    return line == "end" or line.startswith(_MERMAID_LINE_PREFIX_SKIP)


def _normalize_mermaid_label(label: str) -> str:
    cleaned = _MERMAID_LINE_BREAK_RE.sub(" ", label)
    cleaned = _MERMAID_HTML_TAG_RE.sub(" ", cleaned)
    cleaned = cleaned.replace("&nbsp;", " ")
    cleaned = re.sub(r"\s+", " ", cleaned).strip()
    cleaned = cleaned.replace("\\", "\\\\").replace('"', '\\"')
    return cleaned


def _replace_rect_node(match: re.Match[str]) -> str:
    node_id = match.group("id")
    label = _normalize_mermaid_label(match.group("label"))
    return f'{node_id}["{label}"]'


def sanitize_mermaid_diagram(diagram: str) -> str:
    normalized = diagram.replace("\r\n", "\n").replace("\r", "\n").strip()
    normalized = _MERMAID_LINE_BREAK_RE.sub(" ", normalized)
    normalized = normalized.replace("&nbsp;", " ")
    normalized = _MERMAID_GRAPH_RE.sub(r"\1flowchart\2", normalized, count=1)

    sanitized_lines: list[str] = []
    for line in normalized.split("\n"):
        stripped = line.lstrip()
        if _should_skip_mermaid_line(stripped):
            sanitized_lines.append(line.rstrip())
            continue
        sanitized_lines.append(_MERMAID_RECT_NODE_RE.sub(_replace_rect_node, line).rstrip())

    return "\n".join(sanitized_lines).strip()


def sanitize_mermaid_blocks(markdown: str) -> str:
    if "```mermaid" not in markdown:
        return markdown

    def _replace_block(match: re.Match[str]) -> str:
        diagram = sanitize_mermaid_diagram(match.group(1))
        return f"```mermaid\n{diagram}\n```"

    return _MERMAID_FENCE_RE.sub(_replace_block, markdown)
